packages/waypoint/headless/track/track.ts

```diff
@@ -0,0 +1,266 @@
+  headers.set("Accept", CONTENT_TYPE)
+
+  const body = JSON.stringify({
+    events,
+  })
+
+  return fetch(TRACK_URL, {
+    method: "POST",
+    headers,
+    body,
+  })
+}
+
+const TRACKING_OFFSET_KEY = "WAYPOINT.HEADLESS.TRACKING_OFFSET"
+const getOffset = () => {
+  const storageReady = typeof window !== "undefined" && "sessionStorage" in window
+
+  if (!storageReady) return 0
+
+  try {
+    const currentOffset = parseInt(sessionStorage.getItem(TRACKING_OFFSET_KEY) ?? "0")
+    const newOffset = currentOffset + 1
+
+    sessionStorage.setItem(TRACKING_OFFSET_KEY, newOffset.toString())
+    return currentOffset
+  } catch (error) {
+    return 0
+  }
+}
+const getUTCTime = () => {
+  const date = new Date()
+  const timestamp = `${date.getUTCFullYear()}-${date.getUTCMonth()}-${date.getUTCDate()} ${date.getUTCHours()}:${date.getUTCMinutes()}:${date.getUTCSeconds()}`
+
+  return timestamp
+}
+const toErrorProperties = (error: unknown): ErrorProperties => {
+  if (error instanceof HeadlessClientError)
+    return {
+      error_level: "error",
+      error_type: "client_defined",
+      error_name: error.name,
+      error_code: error.code,
+      error_message: error.shortMessage,
+    }
+
+  if (error instanceof ServerError)
+    return {
+      error_level: "error",
+      error_type: "server",
+      error_name: error.name,
+      error_code: error.code,
+      error_message: error.shortMessage,
+    }
+
+  if (error instanceof Error)
+    return {
+      error_level: "error",
+      error_type: "client_unknown",
+      error_name: error.name,
+      error_code: HeadlessClientErrorCode.UnknownError,
+      error_message: error.message,
+    }
+
+  return {
+    error_level: "error",
+    error_type: "client_unknown",
+    error_name: "UnknownError",
+    error_code: HeadlessClientErrorCode.UnknownError,
+    error_message: "Unknown error",
+  }
+}
+
+type CreateTrackerParams = {
+  event: HeadlessEventName
+  waypointToken: string
+
+  wasmUrl?: string
+  productionFactor?: string | boolean
+}
+export const createTracker = (params: CreateTrackerParams) => {
+  const { event, waypointToken, wasmUrl = "", productionFactor = false } = params
+  const startTime = performance.now()
+
+  const _getCommonData = () => {
+    const endTime = performance.now()
+    const duration = endTime - startTime
+    const {
+      iss = "",
+      sub = "",
+      sid = "",
+      aud = [],
+      client_id = "",
+    } = jwtDecode<WaypointTokenPayload & ASAccessTokenPayload>(waypointToken)
+
+    return {
+      event,
+      uuid: crypto.randomUUID(),
+      offset: getOffset(),
+      timestamp: getUTCTime(),
+      session_id: sid,
+      user_id: iss && sub ? sha256(stringToHex(`${iss}:${sub}`)) : "",
+      ref: "",
+
+      commonActionProperties: {
+        aud,
+        iss,
+        sub,
+        client_id,
+
+        origin: window?.location?.origin ?? "",
+        user_agent: navigator?.userAgent ?? "",
```

Comment:
này cũng zị luôn

Reply:
mấy này `?` xong nó có bị gì không

Reply:
ở server nó không có window, nên khi gọi đầu window là nó chửi luôn window is not defined, chứ chưa tới "?." nữa. nên người ta sẽ xét typeof window, lúc này typeof nó cover nên không bị chửi

Reply:
ok sir